feat: add input type support for default values

nickofthyme opened the pull request:
Attempts 1 & 2 - Failed ❌ 

You can pull down this branch and play around with it. I used `src/platform/packages/shared/kbn-config-schema/testing-defaults.ts` as simple testing file.

## Summary

Adds support for `TypeOfInput` and `TypeOfOutput` types on `@kbn/config-schema`

```ts
import { schema, TypeOf, TypeOfOutput, TypeOfInput } from '@kbn/config-schema';

const mySchema = schema.object(
  {
    id: schema.number(),
    maybe: schema.maybe(schema.number()),
    num: schema.number({ defaultValue: 2 }),
    obj: schema.object(
      {
        test1: schema.number(),
        test2: schema.number({ defaultValue: 2 }),
      },
      {
        defaultValue: { test1: 1 },
      }
    ),
  },
);

type MySchemaType = Simplify<TypeOf<typeof mySchema>>; // same as output
type MySchemaOuput =  Simplify<TypeOfOutput<typeof mySchema>>;

// Printed type
type MySchemaType = MySchemaOuput;
type MySchemaOuput = {
    id: number;
    num: number;
    obj: {
        test1: number;
        test2: number;
    };
    maybe?: number | undefined;
};

type MySchemaInput = Simplify<TypeOfInput<typeof mySchema>>;

// Printed type
type MySchemaInput = {
    id: number;
    maybe?: number | undefined;
    num?: number | undefined;
    obj?: {
        test1: number;
        test2?: number | undefined;
    } | undefined;
}
```

## Details

These types are crazy! It takes a lot to pull the `defaultValue` from the object and correctly apply that to reshape the `defaultValue` of the object and pass the inputs correctly while avoiding circular references.

### Checklist

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

nickofthyme committed:
feat: add input type support for default values

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

nickofthyme committed:
Merge branch 'main' into schema-defaults

nickofthyme committed:
fix: default object values should be validated to fill deeper defaults

nickofthyme committed:
fix more types like `oneOf` and `allOf` and others

nickofthyme committed:
chore: fix general type for `TypeOfX` helper types

nickofthyme committed:
fix: intersection type and `allOf` schema types

nickofthyme committed:
fix: `schema.bytesSize` types

nickofthyme committed:
fix: `schema.lazy` type

nickofthyme committed:
fix: contitional type to allow different default values between option `A` and `B`.

nickofthyme committed:
fix: offering based schema

nickofthyme committed:
fix: maybe type

nickofthyme committed:
fix: literal type, automatically defaults to the value

nickofthyme committed:
fix: `nullable` type

nickofthyme committed:
fix: simple type errors

nickofthyme committed:
fix `oneOf` to allow mixing `Type` and `ObjectType`

nickofthyme committed:
other cleanup

nickofthyme committed:
array fixes

nickofthyme committed:
fix arrayOf type

nickofthyme committed:
fix `mapOf` type to allow defaults

nickofthyme committed:
improve `oneOf` performance

nickofthyme committed:
fix refs to preserve type

nickofthyme committed:
fix: typeof types for `oneOf` types

nickofthyme committed:
add type check tests for `TypeOf` types

nickofthyme committed:
improve ref generic typings

nickofthyme committed:
relocate general type tests to respective tests

nickofthyme committed:
add intersection merge type

nickofthyme committed:
cleanup some shared types, add helpers

nickofthyme committed:
fix `allOf` types......again!

nickofthyme committed:
cleanup intersection types

nickofthyme committed:
loosen default value on `Type` and `ObjectType`

These types ideally are the same, if the value is a number I expect the default to be a number. But it gets complicated with objects as the props can now be different depending on the defaults supplied. Additionally the union type can be different than the original value as it also support objects.

nickofthyme committed:
more type fixes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

nickofthyme committed:
rm testing file

nickofthyme committed:
fix duration default type issues

The `schema.duration` allows passing a default as a string or number which is used to generate a Duration. But it also allowed a reference to a string or number which would never be converted to a string. This makes references only be a Duration.

nickofthyme committed:
Merge branch 'main' into schema-defaults

nickofthyme committed:
Merge remote-tracking branch 'origin/schema-defaults' into schema-defaults

nickofthyme commented:
/ci

nickofthyme committed:
fix `schema.object` validate callback option
